### ISSUE TITLE
Pantheon: disable gnome ssh agent when gpg-agent is used for SSH

### ISSUE THREAD

**Issue by kirelagin** [NixOS/nixpkgs#101616]
NixOS has options that allow one to configure `gpg-agent` system-wide: `programs.gnupg.agent.*`. When `enable = true`, it creates a user service that starts `gpg-agent` and, additionally, when `enableSSHSupport = true` it adds shell initialisation code that exports `SSH_AUTH_SOCK` (this should be a user service too, so that it will work on Wayland, but that’s a separate story).

Anyway, I think it is expected that if both `enable` and `enableSSHSupport` are set to `true`, NixOS will setup `gpg-agent` to be used as an SSH agent.

However, Pantheon uses Gnome Keyring, which has an SSH component. This component, among other things, sets `SSH_AUTH_SOCK` to point to the Gnome Keyring’s socket, and this happens relatively late in the session initialisation process to take precedence over all other places where you could export this variable.

The solution that I discovered was to copy `pantheon-gnome-keyring-ssh.desktop` from one of Pantheon packages, add `Hidden=true` to the file, and put it in `$HOME/.config/autostart`, where it shadows the one installed in the system.

Would it make sense to teach the Pantheon NixOS module to drop this file if ssh support is enabled in the gpg agent?

**Comment by worldofpeace** [comment]
> Would it make sense to teach the Pantheon NixOS module to delete this file if ssh support is enabled in the gpg agent?

I believe there is an environment variable that GNOME Session accepts to just "not do this"

**Comment by kirelagin** [comment]
No, that’s different. There used to be a crazy hack in `gnome-session` that just set `SSH_AUTH_SOCK`, unless this special variable you are probably referring to was set. This crazy hack was only for Wayland and is long gone, since they implemented stages in `gnome-session`. What I am talking about is not a hack, but actually Elementary explicitly requesting this to happen.

Here are some details:

1. `elementary-session-settings` copies autostart files from somewhere Gnome-related (I am not smart enough to figure our where from exactly). Among those files [is `gnome-keyring-ssh.desktop`](https://github.com/elementary/session-settings/blob/c6e743fd75b773f5d3afb83288494b7b70b18bf0/session/meson.build#L122).
2. Since `elementary-session-settings` is added to `system.path` by the Pantheon NixOS module, its contents get linked to `/run/current-system/sw`, in particular, `/run/current-system/sw/etc/xdg/autostart/pantheon-gnome-keyring-ssh.desktop`.
3. `/run/current-system/sw/etc/xdg` is added to `$XDG_CONFIG_DIRS`, so `gnome-session` checks its `autostart/` directory for `.desktop` files to start and thus it runs `/run/wrappers/bin/gnome-keyring-daemon --start --components=ssh`.
4. This component of Keychain [sets `SSH_AUTH_SOCK`](https://github.com/GNOME/gnome-keyring/blob/6c19deb36ade13d573bd8e861c2ab7e5f09065ee/daemon/ssh-agent/gkd-ssh-agent-service.c#L394) in the context of ~Gnome~Pantheon session.

There are two ways to disable this behaviour: 1. remove this `.desktop` file; 2. shadow it with another file with `Hidden=true` inside.

**Comment by kisik21** [comment]
This is still important to me. I've tried to shadow the `.desktop` file with `Hidden=true` but it is either flaky or stopped working because I've been encountering this issue again. Having to override this variable manually or even in .bashrc feels hacky. Maybe we could find a way to solve it once and for all?...

**Comment by avitex** [comment]
The [gnome-keyring PAM module](https://github.com/NixOS/nixpkgs/blob/9b962429be9a6820ec1746a0e6350cfe98cc9bcf/nixos/modules/security/pam.nix#L310) gets enabled when the associated [gnome-keyring service](https://github.com/NixOS/nixpkgs/blob/468cb5980b56d348979488a74a9b5de638400160/nixos/modules/services/desktops/gnome/gnome-keyring.nix#L52) is by [one](https://github.com/NixOS/nixpkgs/blob/07978cddabe1ddce172b1376560add398750e2d3/nixos/modules/services/x11/desktop-managers/gnome.nix) [of](https://github.com/NixOS/nixpkgs/blob/c0c34eb75783e1d7b035015d207e5b3d742d26b3/nixos/modules/services/x11/desktop-managers/cinnamon.nix) [the](https://github.com/NixOS/nixpkgs/blob/468cb5980b56d348979488a74a9b5de638400160/nixos/modules/services/x11/desktop-managers/pantheon.nix) [many](https://github.com/NixOS/nixpkgs/blob/468cb5980b56d348979488a74a9b5de638400160/nixos/modules/programs/geary.nix) dependents that enable it. This starts the `gnome-keyring-daemon` which by default has the internal `SSH` component enabled. That component sets the `SSH_AUTH_SOCK` environment variable by both [pushing it](https://gitlab.gnome.org/GNOME/gnome-keyring/-/blob/master/daemon/gkd-glue.c#L61) over [DBUS to gnome-session](https://gitlab.gnome.org/GNOME/gnome-keyring/-/blob/master/daemon/dbus/gkd-dbus-environment.c#L42) and via [`g_setenv`](https://gitlab.gnome.org/GNOME/gnome-keyring/-/blob/master/daemon/ssh-agent/gkd-ssh-agent-service.c#L394).

I have submitted a [PR to gnome-keyring](https://gitlab.gnome.org/GNOME/gnome-keyring/-/merge_requests/40) to allow [specifying the components that should be started in the daemon from the PAM module](https://github.com/avitex/nixpkgs/commit/7e760ee360da32795dfee51e4b0aaf6453a4af38)[1]. This will allow omitting the `SSH` component.

I think this could be a solution going forward without completely breaking things

[1] This configuration was just for testing

**Comment by CMCDragonkai** [comment]
I'm wondering about this too, but not for Pantheon. I don't use a desktop manager at all, and in my case both gnome keyring and gpg agent works fine. But gnome keyring appears to be loading my ssh keys too.

When gnome keyring is enabled, it appears to override the `SSH_ASKPASS` from `/nix/store/dy17b2dkadsawv2sdbk5mak26lrv8y32-x11-ssh-askpass-1.2.4.1/libexec/x11-ssh-askpass` to using gnome keyring instead.

The PR doesn't seem to have been accepted. So I'm guessing right now gnome keyring and gpg agent are both loading the ssh keys?

**Comment by trev-dev** [comment]
This is driving me bonkers.  How can I make it stop?  Avoid Gnome?

**Comment by rgoulter** [comment]
> This is driving me bonkers. How can I make it stop? Avoid Gnome?

https://wiki.archlinux.org/title/GNOME/Keyring#Disable_keyring_daemon_components

This worked for me.

Of course, `gnome-keyring-ssh.desktop` isn't under `/etc` in NixOS; so you'll have to go to `/nix/store` and find with e.g. `find . -name gnome-keyring-ssh.desktop 2>/dev/null`.

**Comment by avitex** [comment]
> The [gnome-keyring PAM module](https://github.com/NixOS/nixpkgs/blob/9b962429be9a6820ec1746a0e6350cfe98cc9bcf/nixos/modules/security/pam.nix#L310) gets enabled when the associated [gnome-keyring service](https://github.com/NixOS/nixpkgs/blob/468cb5980b56d348979488a74a9b5de638400160/nixos/modules/services/desktops/gnome/gnome-keyring.nix#L52) is by [one](https://github.com/NixOS/nixpkgs/blob/07978cddabe1ddce172b1376560add398750e2d3/nixos/modules/services/x11/desktop-managers/gnome.nix) [of](https://github.com/NixOS/nixpkgs/blob/c0c34eb75783e1d7b035015d207e5b3d742d26b3/nixos/modules/services/x11/desktop-managers/cinnamon.nix) [the](https://github.com/NixOS/nixpkgs/blob/468cb5980b56d348979488a74a9b5de638400160/nixos/modules/services/x11/desktop-managers/pantheon.nix) [many](https://github.com/NixOS/nixpkgs/blob/468cb5980b56d348979488a74a9b5de638400160/nixos/modules/programs/geary.nix) dependents that enable it. This starts the `gnome-keyring-daemon` which by default has the internal `SSH` component enabled. That component sets the `SSH_AUTH_SOCK` environment variable by both [pushing it](https://gitlab.gnome.org/GNOME/gnome-keyring/-/blob/master/daemon/gkd-glue.c#L61) over [DBUS to gnome-session](https://gitlab.gnome.org/GNOME/gnome-keyring/-/blob/master/daemon/dbus/gkd-dbus-environment.c#L42)[1] and via [`g_setenv`](https://gitlab.gnome.org/GNOME/gnome-keyring/-/blob/master/daemon/ssh-agent/gkd-ssh-agent-service.c#L394).
> 
> I have submitted a [PR to gnome-keyring](https://gitlab.gnome.org/GNOME/gnome-keyring/-/merge_requests/40) to allow [specifying the components that should be started in the daemon from the PAM module](https://github.com/avitex/nixpkgs/commit/7e760ee360da32795dfee51e4b0aaf6453a4af38)[2]. This will allow omitting the `SSH` component.
> 
> I think this could be part of the solution going forward without completely breaking things
> 
> [1] [org.gnome.SessionManager.Setenv](https://people.gnome.org/~mccann/gnome-session/docs/gnome-session.html#org.gnome.SessionManager.Setenv) [2] This configuration was just for testing

In respect to this issue, I've drafted a PR to disable the SSH component that's enabled in the gnome-keyring PAM module if `programs.gnupg.agent.enableSSHSupport == true`. See https://github.com/NixOS/nixpkgs/pull/284173
